### PR TITLE
bpf: Separate conntrack timeouts for TCP and non-TCP

### PR DESCRIPTION
**Pull request by tgraf**
The current conntrack expire behavior is as follows:
 - 5 minute reduced lifetime for TCP entries when only SYN has been seen
 - 12 hour lifetime for entries when a packet other than a SYN has been seen
 - Removal of entries after observing bidirectional FIN
 - Removal of entries on policy rejection

This behavior works well for TCP but leaves the conntrack table exposed to UDP
workloads which generate a lot of unique requests.

Make the conntrack expiration logic aware of UDP and apply a reduced 5 minute
lifetime for conntrack entries. Also reduce the lifetime of TCP established
connections to 6 hours as all L7 proxies have been configured to apply TCP
keep alive behavior.

While at it, make the expiration times configurable via the header files to
allow introducing configurable timeouts in a later commit.

Signed-off-by: Thomas Graf <thomas@cilium.io>

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/4389)

### CODE REVIEW

**Comment by tgraf** [comment]
test-me-please

**Review by joestringer — APPROVED** [review]
LGTM. I was under the impression that Linux limits non-TCP to timeouts more like 30 or 60 seconds, 5 mins seems high to me. Seems still quite easy to flood the CT table - ~3.3K UDP requests per second over 5 minutes.

**Comment by tgraf** [comment]
> LGTM. I was under the impression that Linux limits non-TCP to timeouts more like 30 or 60 seconds, 5 mins seems high to me. Seems still quite easy to flood the CT table to me - ~3.3K UDP requests per second over 5 minutes.

Yes, I had the same thought. I'll drop it down to 60 seconds.

**Comment by joestringer** [comment]
Hmm, with the default options in the `bpf/` directory we're seeing this in the testsuite:

```
	 BPF program is too large. Proccessed 65537 insn
```

**Comment by tgraf** [comment]
test-me-please

**Comment by tgraf** [comment]
> BPF program is too large. Proccessed 65537 insn

I've rearranged the code a bit and was able to generate endpoints in my vagrant environment. Let's see how it goes in the CI environment.

**Comment by joestringer** [comment]
@tgraf hit the same issue in the CI. I'd recommend either manually running `test/bpf/verifier_test.sh` or running it via the ginkgo test, `RuntimeValidatedVerifier`.

We can always just reduce the number of CIDR prefix lengths we have in the `bpf/lxc_config.h` and it will address this problem in the build environment, but we should be aware that these default headers in the `bpf/` directory are trying to give an indicator for the maximum complexity we may generate at runtime.

**Comment by tgraf** [comment]
test-me-please

**Comment by aanm** [comment]
@tgraf does this needs backport?

**Comment by tgraf** [comment]
> does this needs backport?

Yes

**Comment by aanm** [comment]
@tgraf I've fixed the labels, I assume you wanted `needs-backport` and not `backport`

File: pkg/maps/ipcache/ipcache.go [hunk 1]
```diff
@@ -41,7 +41,7 @@ const (
 	// (/8) will always be inserted, which is what the first parameter
 	// denotes. The CIDR for the CIDRSet is the second parameter, and the
 	// exception is the third parameter.
-	maxPrefixLengths = 2 - 8 + 32
+	maxPrefixLengths = 2 - 8 + 26
```
Comment (joestringer) [inline pkg/maps/ipcache/ipcache.go:44]: This is actually now restricting us to 20 unique prefixes, while the `lxc_config.h` is stating that we should support ~26.

**Comment by tgraf** [comment]
I've sunstracted 6 from both values. It must have been out of sync before

**Comment by joestringer** [comment]
Yes, the `maxPrefixLengths` wasn't generated from the `lxc_config.h` before. The comment above the variable describes where it did come from. We can simplify the definition of this variable to just match what we put in `lxc_config.h`.

**Pull request merged**